plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/FileSystemExchange.java

```diff
@@ -279,9 +280,19 @@ private URI getTaskOutputDirectory(int taskPartitionId)
     }
 
     @Override
-    public CompletableFuture<List<ExchangeSourceHandle>> getSourceHandles()
+    public ExchangeSourceHandleSource getSourceHandles()
     {
-        return exchangeSourceHandlesFuture;
+        return new ExchangeSourceHandleSource()
+        {
+            @Override
+            public CompletableFuture<ExchangeSourceHandleBatch> getNextBatch()
+            {
+                return exchangeSourceHandlesFuture.thenApply(handles -> new ExchangeSourceHandleBatch(handles, true));
+            }
+
+            @Override
+            public void close() {}
```

Comment:
should this cancel the future?

Reply:
The future is a singleton. If we want to allow exchange sources to be enumerated more than once the underlying future should not be cancelled.